flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/sink/writer/JdbcWriterState.java

```diff
@@ -0,0 +1,105 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.jdbc.sink.writer;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.connector.jdbc.datasource.transactions.xa.domain.TransactionId;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.apache.commons.lang3.builder.ToStringStyle;
+
+import javax.annotation.concurrent.ThreadSafe;
+import javax.transaction.xa.Xid;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+
+/** Thread-safe (assuming immutable {@link Xid} implementation). */
+@ThreadSafe
+@Internal
+public class JdbcWriterState implements Serializable {
+    private final Collection<TransactionId> prepared;
+    private final Collection<TransactionId> hanging;
+
+    public static JdbcWriterState empty() {
+        return new JdbcWriterState(Collections.emptyList(), Collections.emptyList());
+    }
+
+    public static JdbcWriterState of(
+            Collection<TransactionId> prepared, Collection<TransactionId> hanging) {
+        return new JdbcWriterState(
+                Collections.unmodifiableList(new ArrayList<>(prepared)),
+                Collections.unmodifiableList(new ArrayList<>(hanging)));
+    }
+
+    protected JdbcWriterState(
+            Collection<TransactionId> prepared, Collection<TransactionId> hanging) {
+        this.prepared = prepared;
+        this.hanging = hanging;
+    }
+
+    /**
+     * @return immutable collection of prepared XA transactions to {@link
+     *     javax.transaction.xa.XAResource#commit commit}.
+     */
+    public Collection<TransactionId> getPrepared() {
+        return prepared;
+    }
+
+    /**
+     * @return immutable collection of XA transactions to {@link
+     *     javax.transaction.xa.XAResource#rollback rollback} (if they were prepared) or {@link
+     *     javax.transaction.xa.XAResource#end end} (if they were only started).
+     */
+    public Collection<TransactionId> getHanging() {
+        return hanging;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        JdbcWriterState that = (JdbcWriterState) o;
+        return new EqualsBuilder()
+                .append(prepared, that.prepared)
+                .append(hanging, that.hanging)
+                .isEquals();
```

Comment:
is there no way to use just `equals` ?
It seems collections only, so equals e.g. `Objects.equals` should work...
I'm confused by creation a new object each time `equals` is called.
Or did I miss anything?

Reply:
is auto generated, with template of apache commons-lang3

I could change it if the default is another way